### ISSUE TITLE
iOS keeps using cellular data after default route changes to WiFi

### ISSUE THREAD

**Issue by mihaip** [tailscale/tailscale#6565]
@dsnet reported that when his iPhone gets on WiFi it continues to communicate with his NAS over the cellular connection. We investigated, and got as far as noticing that Magicsock tries to do DISCO pings over the new endpoints that it was told about, but bi-directional communication is not possible:

These are the [logs from Joe’s phone](http://logz.ts.tailscale.com:9431/logs?start=2022-11-10T22%3A19%3A57Z&end=now&instances=d25cc550108d0dad040aa78f4dd581fad1a83d244ec43c6f0e4552e66c7b9edd), you can see that it both sends and gets a ping from his NAS over his LAN (192.168.0.101):

```
magicsock: disco: d:baf1be2bc22288d6->d:63f0e05d66bb491d ([K94YI], 192.168.0.101:41641) sent ping tx=8609cd4c7a33
magicsock: disco: d:baf1be2bc22288d6<-d:63f0e05d66bb491d ([K94YI], 192.168.0.101:41641) got ping tx=076ceb0acb7a
```

The [NAS’s logs](http://logz.ts.tailscale.com:9431/logs?start=2022-11-10T22%3A19%3A57Z&end=now&instances=afad3037420ff1d02cf2b22eeae5fee2c476b9912fc0d9fbd41fabfe9f1d9f2e) show the ping that it sent, but it doesn’t get the ping or pong from his phone:

```
magicsock: disco: d:63f0e05d66bb491d->d:baf1be2bc22288d6 ([wNcRJ], 192.168.0.183:41641) sent ping tx=076ceb0acb7a
```

We’re assuming that the iOS network stack is trying to send the UDP packets to 192.168.0.101 over the cell connection. It would be nice if when we discover new interfaces/endpoints we bind the sockets to those interfaces instead of letting iOS choose.

_Originally posted by @mihaip in https://github.com/tailscale/tailscale/issues/6152#issuecomment-1315910954_

**Comment by mihaip** [comment]
I did some more investigating into the issue that Joe saw -- I can reproduce it when having my iPhone connect to my Synology (LAN IP `192.168.1.100`) -- it keeps using the cell connection even after WiFi is available. I'm guessing it's a race between when `darwinRouteMon` notifies of the link change and when iOS truly thinks that the new interface can be used. 

Here's some log messages showing that the iOS-side network change notification (the `from Swift` one) is received ~100ms later:
```
default	15:28:37.874860-0800	IPNExtension	monitor interfaces old: interfaces.State{defaultRoute=pdp_ip0 ifs={en0:[192.168.1.124/24 fd3f:46a3:a86:47e9:d9:3554:6a8c:e91f/64] ipsec3:[fd74:6572:6d6e:7573:c:44b2:79d4:eb0b/128 fd74:6572:6d6e:7573:d:44b2:79d4:eb0b/128] ipsec4:[192.0.0.6/32 fd74:6572:6d6e:7573:c:44b2:79d4:eb0b/64 fd74:6572:6d6e:7573:d:44b2:79d4:eb0b/64] pdp_ip0:[100.115.1.209/32 2600:1010:b05f:bb1f:14fc:c9d9:c61:f613/64 2600:1010:b05f:bb1f:8dcc:fa3d:881d:c458/64] pdp_ip1:[2600:1010:1038:b65c:c54:4c97:35df:627d/64 2600:1010:1038:b65c:e95b:a10d:604d:570b/64] utun23:[100.113.218.14/32 fd7a:115c:a1e0:ab12:4843:cd96:6271:da0e/48]} v4=true v6=true}
default	15:28:37.874953-0800	IPNExtension	monitor interfaces new: interfaces.State{defaultRoute=en0 ifs={en0:[192.168.1.124/24 fd3f:46a3:a86:47e9:d9:3554:6a8c:e91f/64] ipsec3:[fd74:6572:6d6e:7573:c:44b2:79d4:eb0b/128 fd74:6572:6d6e:7573:d:44b2:79d4:eb0b/128] ipsec4:[192.0.0.6/32 fd74:6572:6d6e:7573:c:44b2:79d4:eb0b/64 fd74:6572:6d6e:7573:d:44b2:79d4:eb0b/64] pdp_ip0:[100.115.1.209/32 2600:1010:b05f:bb1f:14fc:c9d9:c61:f613/64 2600:1010:b05f:bb1f:8dcc:fa3d:881d:c458/64] pdp_ip1:[2600:1010:1038:b65c:c54:4c97:35df:627d/64 2600:1010:1038:b65c:e95b:a10d:604d:570b/64] utun23:[100.113.218.14/32 fd7a:115c:a1e0:ab12:4843:cd96:6271:da0e/48]} v4=true v6=true}
default	15:28:37.875116-0800	IPNExtension	monitor interfaces changed
default	15:28:37.876358-0800	IPNExtension	Rebind; defIf="en0", ips=[192.168.1.124/24 fd3f:46a3:a86:47e9:d9:3554:6a8c:e91f/64 fe80::18ea:2c55:3563:5125/64]
default	15:28:37.922138-0800	IPNExtension	magicsock: disco: d:fbf3ecd696af491c->d:5cacb9dac8f03232 ([PQvTs], 192.168.1.100:41641) sent ping tx=b96d3ea06ae2
default	15:28:37.984390-0800	IPNExtension	magicsock: disco: d:fbf3ecd696af491c<-d:5cacb9dac8f03232 ([PQvTs], 192.168.1.100:41641)  got ping tx=cdfbd8ddb3d6
default	15:28:37.997322-0800	IPNExtension	network update from Swift (ignored): connected=true, expensive=false
```
(no full roundtrips with `192.168.1.100` are ever exchanged)

If I inject an 1 second delay into the rebinding of the UDP connections at https://github.com/tailscale/tailscale/blob/dd50dcd06768bc54112a9acc5229d9b9ef63832c/wgengine/userspace.go#L1214 then magicsock is able to successfully do the end-to-end ping pong and switches over:

```
default	18:21:46.092020-0800	IPNExtension	monitor interfaces old: interfaces.State{defaultRoute=pdp_ip0 ifs={en0:[192.168.1.124/24 fd3f:46a3:a86:47e9:d9:3554:6a8c:e91f/64] ipsec3:[fd74:6572:6d6e:7573:c:44b2:79d4:eb0b/128 fd74:6572:6d6e:7573:d:44b2:79d4:eb0b/128] ipsec4:[192.0.0.6/32 fd74:6572:6d6e:7573:c:44b2:79d4:eb0b/64 fd74:6572:6d6e:7573:d:44b2:79d4:eb0b/64] pdp_ip0:[100.115.1.209/32 2600:1010:b05f:bb1f:14fc:c9d9:c61:f613/64 2600:1010:b05f:bb1f:8dcc:fa3d:881d:c458/64] pdp_ip1:[2600:1010:1038:b65c:c54:4c97:35df:627d/64 2600:1010:1038:b65c:e95b:a10d:604d:570b/64] utun29:[100.113.218.14/32 fd7a:115c:a1e0:ab12:4843:cd96:6271:da0e/48]} v4=true v6=true}
default	18:21:46.092072-0800	IPNExtension	monitor interfaces new: interfaces.State{defaultRoute=en0 ifs={en0:[192.168.1.124/24 fd3f:46a3:a86:47e9:d9:3554:6a8c:e91f/64] ipsec3:[fd74:6572:6d6e:7573:c:44b2:79d4:eb0b/128 fd74:6572:6d6e:7573:d:44b2:79d4:eb0b/128] ipsec4:[192.0.0.6/32 fd74:6572:6d6e:7573:c:44b2:79d4:eb0b/64 fd74:6572:6d6e:7573:d:44b2:79d4:eb0b/64] pdp_ip0:[100.115.1.209/32 2600:1010:b05f:bb1f:14fc:c9d9:c61:f613/64 2600:1010:b05f:bb1f:8dcc:fa3d:881d:c458/64] pdp_ip1:[2600:1010:1038:b65c:c54:4c97:35df:627d/64 2600:1010:1038:b65c:e95b:a10d:604d:570b/64] utun29:[100.113.218.14/32 fd7a:115c:a1e0:ab12:4843:cd96:6271:da0e/48]} v4=true v6=true}
default	18:21:46.092102-0800	IPNExtension	monitor interfaces changed changed
default	18:21:46.092284-0800	IPNExtension	wgengine: major link change, pausing
default	18:21:46.199424-0800	IPNExtension	network update from Swift (ignored): connected=true, expensive=false
default	18:21:47.065444-0800	IPNExtension	wgengine: major link change, resuming
default	18:21:47.065833-0800	IPNExtension	Rebind; defIf="en0", ips=[192.168.1.124/24 fd3f:46a3:a86:47e9:d9:3554:6a8c:e91f/64 fe80::18ea:2c55:3563:5125/64]
default	18:21:47.101982-0800	IPNExtension	magicsock: netInfo update: NetInfo{varies=false hairpin=false ipv6=false ipv6os=true udp=true icmpv4=false derp=#2 portmap= link=""}
default	18:21:48.671257-0800	IPNExtension	magicsock: disco: d:79b3f8b9e67c95eb->d:5cacb9dac8f03232 ([PQvTs], 192.168.1.100:41641) sent ping tx=89474fb9ba77
default	18:21:48.682568-0800	IPNExtension	magicsock: disco: d:79b3f8b9e67c95eb<-d:5cacb9dac8f03232 ([PQvTs], 192.168.1.100:41641)  got pong tx=89474fb9ba77 latency=12ms pong.src=192.168.1.124:41641
default	18:21:48.682667-0800	IPNExtension	magicsock: disco: node [PQvTs] d:5cacb9dac8f03232 now using 192.168.1.100:41641
default	18:21:48.701546-0800	IPNExtension	magicsock: disco: d:79b3f8b9e67c95eb<-d:5cacb9dac8f03232 ([PQvTs], 192.168.1.100:41641)  got ping tx=2115152bc1de
```

Not saying that injecting the delay is what we should actually do, but it does show that there's a timing/race issue. Will look into what kinds of messages `darwinRouteMon` gets and see if we can be better about when we notify (14dc79013754fe40dcda89f8a3394b4d865629a2 mentions actually doing something with the `AF_ROUTE` messages instead of triggering a re-poll of interfaces, it looks like that's still not done)

_Originally posted by @mihaip in https://github.com/tailscale/tailscale/issues/6152#issuecomment-1320757851_

**Comment by isaackwan** [comment]
Thanks for your amazing work on this (PR #6566) finally addressing this long-standing issue. I don't know if an additional data point is useful at this point, but my gf has just burned through 30GB of data overnight. Thankfully no overage charge is levied.

I used her phone to log a bug report: BUG-6e06f0b3e0d8828805b81a6ed645be99833523ec5bc497b973284439e4d6e0ec-20230107024043Z-810bc7e316a3abe0

If you don't mind, can I ask a few questions:
1) I see the PR is already merged, I assume I'll make it to the next stable release 1.36? What is Tailscale's release schedule like, do we have any ballpark figure when might that be?

2) To prevent this from happening again, is there a possibility of introducing a switch, "unmeteted wi-fi only" so no mobile data is used ever? Might help with certain scenarios where we only want to say start syncing data with my NAS when I'm connected to a Wi-Fi.

3) The extent of the previous bug/behaviour, ie potentially using cellular data even when connected to Wi-Fi, does it affect non-Tailscale routes as well? Ie if my client has exit node turned off and surf Facebook, is it using wi-fi or cellular data? I assume cellular (when the bug kicks in, that is)?

**Comment by mihaip** [comment]
>  but my gf has just burned through 30GB of data overnight.

Out of curiosity, how did you determine that Tailscale was the cause? Did it show up in the "Cellular data" settings pane?

> I see the PR is already merged, I assume I'll make it to the next stable release 1.36? What is Tailscale's release schedule like, do we have any ballpark figure when might that be?

Yes, it is currently slated to be picked up in the 1.36 stable release, which should happen in the next couple of weeks. In meantime, if you'd like to try an unstable build with it, you can try it via TestFlight: https://testflight.apple.com/join/tLcYLZJV

> * To prevent this from happening again, is there a possibility of introducing a switch, "unmeteted wi-fi only" so no mobile data is used ever? Might help with certain scenarios where we only want to say start syncing data with my NAS when I'm connected to a Wi-Fi.

We'll consider this, but it would add quite a bit of complexity to the UI and implementation.

> * The extent of the previous bug/behaviour, ie potentially using cellular data even when connected to Wi-Fi, does it affect non-Tailscale routes as well? Ie if my client has exit node turned off and surf Facebook, is it using wi-fi or cellular data? I assume cellular (when the bug kicks in, that is)?

The bug only affected Tailscale connections (either to other Tailscale nodes, or when using an exit nodes). Traffic that would normally not go over Tailscale should behave normally (i.e. use whatever heuristics iOS has for deciding which interface to use -- with things like Wi-Fi assist iOS may choose to use cellular data even when on Wi-Fi).

**Comment by isaackwan** [comment]
Sorry for my slow response. Really appreciate that you are coming back to revisit this issue despite having the PR merged already!

> Out of curiosity, how did you determine that Tailscale was the cause? Did it show up in the "Cellular data" settings pane?
Ha! I'd be happy to be proven wrong. Essentially all her monthly data allowance used up overnight while I was trying to back up her photos using Synology Photos overnight on Wi-Fi. We received text messages by the carrier and we can see that at least 10GB of data is used over night. Synology Photos is set to back up over Wi-Fi only.
And no, I did check the cellular data usage page, but I don't think it showed such huge usage from either Tailscale or Synology Photos.

**Comment by mihaip** [comment]
> And no, I did check the cellular data usage page, but I don't think it showed such huge usage from either Tailscale or Synology Photos.

At least in my experience, if Tailscale is transferring the data,  it will show up for both the Tailscale app and the app that i using it. This is what I see when I use the Synology DS File app to transfer some data over cellular and Tailscale:

![IMG_5079](https://user-images.githubusercontent.com/513813/213520295-90cf521c-ef4d-4daa-b25f-b46667899ad7.jpg)

(the symmetry of the two numbers makes me think that iOS is double-counting something)

**Comment by schaafjs** [comment]
> 2. To prevent this from happening again, is there a possibility of introducing a switch, "unmeteted wi-fi only" so no mobile data is used ever? Might help with certain scenarios where we only want to say start syncing data with my NAS when I'm connected to a Wi-Fi.

If I understand your question correctly, you can already achieve this.
Take a look at "Set cellular data use for apps and services" under https://support.apple.com/guide/iphone/iph3dd5f213/ios.

Note that this may clash with Wi-Fi Assist as noted below the instructions.

**Comment by mihaip** [comment]
Closing this as resolved now that the fix has been picked up in the [1.36 stable release](https://tailscale.com/changelog/#2023-01-24-client).
